File: sdks/rust/examples/transactions.rs
//! Transaction example for Nexus Rust SDK

use nexus_sdk::*;
use std::collections::HashMap;

#[tokio::main]
async fn main() -> Result<()> {
    let client = NexusClient::new("http://localhost:15474")?;

    tracing::info!("=== Transaction Example ===\n");

    // Begin a transaction
    tracing::info!("1. Beginning transaction...");
    let mut tx = client.begin_transaction().await?;
    tracing::info!("   Transaction started");

    // Create nodes within transaction
    tracing::info!("\n2. Creating nodes within transaction...");
    let mut properties1 = HashMap::new();
    properties1.insert("name".to_string(), Value::String("Alice".to_string()));
    let node1 = tx
        .execute("CREATE (n:Person {name: $name}) RETURN id(n) as id", {
            let mut params = HashMap::new();
            params.insert("name".to_string(), Value::String("Alice".to_string()));
            Some(params)
        })
        .await?;
    tracing::info!("   Created node: {:?}", node1);

    let mut properties2 = HashMap::new();
    properties2.insert("name".to_string(), Value::String("Bob".to_string()));
    let node2 = tx
        .execute("CREATE (n:Person {name: $name}) RETURN id(n) as id", {
            let mut params = HashMap::new();
            params.insert("name".to_string(), Value::String("Bob".to_string()));
            Some(params)
        })
        .await?;
    tracing::info!("   Created node: {:?}", node2);

    // Create relationship within transaction
    tracing::info!("\n3. Creating relationship within transaction...");
    let rel = tx
        .execute(
            "MATCH (a:Person {name: 'Alice'}), (b:Person {name: 'Bob'}) CREATE (a)-[r:KNOWS {since: 2020}]->(b) RETURN id(r) as id",
            None,
        )
        .await?;
    tracing::info!("   Created relationship: {:?}", rel);

    // Commit transaction
    tracing::info!("\n4. Committing transaction...");
    tx.commit().await?;
    tracing::info!("   Transaction committed successfully");

    tracing::info!("\n=== Example completed ===");
    Ok(())
}

File: sdks/python/examples/transactions.py
"""Transaction examples for Nexus Python SDK."""

import asyncio
from nexus_sdk import NexusClient, Transaction


async def main():
    """Run transaction examples."""
    async with NexusClient("http://localhost:15474") as client:
        # Begin a transaction using Transaction class
        print("=== Beginning Transaction ===")
        tx = await client.begin_transaction()
        print(f"Transaction ID: {tx.transaction_id}")
        print(f"Transaction active: {tx.is_active()}")
        print(f"Transaction status: {tx.status()}\n")

        # Execute queries within transaction
        print("=== Creating Nodes in Transaction ===")
        result1 = await tx.execute(
            "CREATE (n:Person {name: $name}) RETURN n",
            {"name": "Alice"},
        )
        print(f"Created node 1: {result1.rows}\n")

        result2 = await tx.execute(
            "CREATE (n:Person {name: $name}) RETURN n",
            {"name": "Bob"},
        )
        print(f"Created node 2: {result2.rows}\n")

        # Commit transaction
        print("=== Committing Transaction ===")
        await tx.commit()
        print(f"Transaction status after commit: {tx.status()}\n")

        # Example: Rollback on error
        print("=== Example: Rollback on Error ===")
        tx2 = await client.begin_transaction()
        try:
            # Try to execute a query
            await tx2.execute("CREATE (n:Person {name: 'Charlie'}) RETURN n", None)
            # If we get here, commit
            await tx2.commit()
            print("Transaction committed")
        except Exception as e:
            print(f"Error occurred: {e}")
            # Rollback on error
            await tx2.rollback()
            print(f"Transaction rolled back. Status: {tx2.status()}")


if __name__ == "__main__":
    asyncio.run(main())

